assets/src/asset.rs

```diff
@@ -0,0 +1,165 @@
+///
+/// * the extension (the format it was provided in)
+/// * it's name
+/// * the storage it was loaded from
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct AssetSpec {
+    /// The extension of this asset
+    pub ext: &'static str,
+    /// The name of this asset.
+    pub name: String,
+    /// The storage id, uniquely identifying the storage it's been loaded from.
+    pub store: StoreId,
+}
+
+impl AssetSpec {
+    /// Creates a new asset specifier from the given parameters.
+    pub fn new(name: String, ext: &'static str, store: StoreId) -> Self {
+        AssetSpec {
+            ext,
+            name: name,
+            store,
+        }
+    }
+}
+
+/// A basic implementation for a cache. This might be useful as the `Context` of
+/// an `Asset`, so that the same asset doesn't get imported twice.
+///
+/// Because contexts have to be immutable, a `RwLock` is used. Therefore, all
+/// operations are blocking (but shouldn't block for a long time).
+pub struct Cache<T> {
+    map: RwLock<FnvHashMap<AssetSpec, T>>,
+}
+
+impl<T> Cache<T>
+    where T: Clone
+{
+    /// Creates a new `Cache` and initializes it with the default values.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Inserts an asset, locking the internal `RwLock` to get write access to the hash map.
+    pub fn insert(&self, spec: AssetSpec, asset: T) {
+        self.map.write().insert(spec, asset);
+    }
+
+    /// Retrieves an asset, locking the internal `RwLock` to get read access to the hash map.
+    /// In case this asset has been inserted previously, it will be cloned and returned.
+    /// Otherwise, you'll `None`.
+    pub fn get(&self, spec: &AssetSpec) -> Option<T> {
+        self.map.read().get(spec).map(Clone::clone)
+    }
+
+    /// Deletes all cached values, except the ones `f` returned `true` for.
+    /// May be used when you're about to clear unused assets (see `Asset::clear`).
+    ///
+    /// Blocks the calling thread for getting write access to the hash map.
+    pub fn retain<F>(&self, f: F)
+        where F: FnMut(&AssetSpec, &mut T) -> bool
+    {
+        self.map.write().retain(f);
+    }
+
+    /// Deletes all cached values after locking the `RwLock`.
+    pub fn clear_all(&self) {
+        self.map.write().clear();
+    }
+}
+
+impl<T> Default for Cache<T> {
+    fn default() -> Self {
+        Cache { map: Default::default() }
+    }
+}
+
+/// A format, providing a conversion from bytes to asset data, which is then
+/// in turn accepted by `Asset::from_data`. Examples for formats are
+/// `Png`, `Obj` and `Wave`.
+pub trait Format
+    where Self: Sized
+{
+    /// The data type this format is able to load.
+    type Data;
+    /// The kind of error it may produce.
+    type Error: Error;
+
+    /// Returns the extension (without `.`).
+    ///
+    /// ## Examples
+    ///
+    /// * `"png"`
+    /// * `"obj"`
+    /// * `"wav"`
+    fn extension() -> &'static str;
```

Comment:
I think we will have to mention this behavior (recognizing only one extension per format) in the documentation, because it can potentially be very frustrating to have your `.m4a` file  fail to load while the only thing you have to do is rename it into a `.mp4` file if you don't know about this quirk of the engine.